### PR TITLE
Delete and Abandon Block

### PR DESCRIPTION
**Pull request by Swyrough**

### CODE REVIEW

File: zookeeper/source/zk_dn_client.cc [hunk 1]
```diff
@@ -318,8 +266,44 @@ namespace zkclient{
 		}
 	}
 
-	void ZkClientDn::thisDNDeleteQueueWatcher(zhandle_t *zzh, int type, int state, const char *path, void *watcherCtx){
-		LOG(INFO) << "Delete watcher triggered on path: " << path;
+	void ZkClientDn::processDeleteQueue(std::string path) {
```
Comment (nichhk) [inline zookeeper/source/zk_dn_client.cc]: This doesn't need to take an argument (it's always gonna be the same thing)

File: zookeeper/source/zk_dn_client.cc [hunk 2]
```diff
@@ -94,52 +94,12 @@ namespace zkclient{
 		return created_correctly;
 	}
 
-	bool ZkClientDn::blockDeleted(uint64_t uuid) {
```
Comment (nichhk) [inline zookeeper/source/zk_dn_client.cc:97]: Why get rid of this? We still need to update block locations, block set to show that this datanode no longer has the block.

File: zookeeper/source/zk_nn_client.cc [hunk 3]
```diff
@@ -296,6 +296,105 @@ namespace zkclient{
...
+            return false;
+        }
+        read_file_znode(znode_data, file_path);
+        if (znode_data.filetype != IS_FILE) { // Assert that the znode we want to modify is a file
+            LOG(ERROR) << CLASS_NAME << "Requested file " << file_path << " is not a file";
+            return false;
+        }
+        LOG(INFO) << CLASS_NAME << "File exists. Building multi op to abandon block";
+
+        // Find the last block
```
Comment (nichhk) [inline zookeeper/source/zk_nn_client.cc:347]: We don't need to find the last block. It should be given in the proto.

File: zookeeper/source/zk_nn_client.cc [hunk 4]
```diff
@@ -389,9 +526,10 @@ namespace zkclient{
                 for (auto& dn : datanodes){
                     auto delete_queue = util::concat_path(DELETE_QUEUES, dn);
                     auto delete_item = util::concat_path(delete_queue, "block-");
-                    // TODO: saving this until DataNode team creates delete work queues
-                    //ops.push_back(zk->build_create_op(delete_item, block_vec, ZOO_SEQUENCE));
+                    ops.push_back(zk->build_create_op(delete_item, block_vec, ZOO_SEQUENCE));
+                    blockDeleted(block, dn);
```
Comment (nichhk) [inline zookeeper/source/zk_nn_client.cc:532]: We want the datanode to report block deletion.

File: zookeeper/source/zk_nn_client.cc [hunk 5]
```diff
@@ -1300,6 +1441,27 @@ namespace zkclient{
 		}
 	}
 
+
+    bool ZkNnClient::find_all_datanodes_with_block(const std::string &block_uuid_str, std::vector<std::string>
+        &rdatanodes, int &error_code) {
+
+        std::vector<std::string> datanodes;
```
Comment (nichhk) [inline zookeeper/source/zk_nn_client.cc]: don't need to have separate rdatanodes and datanodes.

File: zookeeper/source/zk_nn_client.cc [hunk 6]
```diff
@@ -342,6 +441,44 @@ namespace zkclient{
...
+
+        // Delete block locations
+        if (zk->exists(BLOCK_LOCATIONS + std::to_string(uuid) + "/" + id, exists, error_code)) {
+            if (exists) {
+                ops.push_back(zk->build_delete_op(BLOCK_LOCATIONS + std::to_string(uuid) + "/" + id));
+                // If deleted last child of block locations, delete block locations
+                std::vector <std::string> children = std::vector <std::string>();
+                if(!zk->get_children(BLOCK_LOCATIONS + std::to_string(uuid), children, error_code)){
+                    LOG(ERROR) << "getting children failed";
+                }
```
Comment (nichhk) [inline zookeeper/source/zk_nn_client.cc:458]: If this DataNode is the only thing with the block, we should clean up BLOCK_LOCATIONS + std::to_string(uuid) too

**Comment by Vegetable26** [comment]
Is there a end-end integration test for this? I think it'd be cool merging this in, if ReplicationTest is untouched

**Comment by Swyrough** [comment]
idk how to respond to ur comment @Vegetable26 . once upon a time there was an integration test, this most recent commit involved me making a multiblock file on two datanodes and then deleting the file and checking for all the proper things myself. ReplicationTest is untouched. Deletes and abandon blocks are 'supplementary' i guess...no demo depends on them.

**Comment by nichhk** [comment]
LGTM

**Comment by Vegetable26** [comment]
I guess I was saying if ReplicationTest still works we should push :) It would be nice to have a test where we write a file and delete it 10 times. Just to make sure all the meta information is gone. And if we used a large file ~500 MB, to see if the DataNode is actually freeing up disk space. But in the sake of going fast, let's just merge 👍

**Pull request merged**